org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Scanner.java

```diff
@@ -1395,7 +1395,8 @@ public int scanIdentifier() throws InvalidInputException {
 }
 public void ungetToken(int unambiguousToken) {
 	if (this.nextToken != TokenNameNotAToken) {
-		throw new ArrayIndexOutOfBoundsException("Single cell array overflow"); //$NON-NLS-1$
+		return;
+		// throw new ArrayIndexOutOfBoundsException("Single cell array overflow"); //$NON-NLS-1$
```

Comment:
I'm not quite comfortable with this change. I can only assume that the exception is thrown because that situation _should never happen_, and we want to be alerted, so we can fix this some other way.

The conflict seems to arise because `fallBackToSpringForward()` is invoked twice without making real progreess. The first invocation sets `this.currentToken = TokenNameSEMICOLON` and the second invocation tries to push that exact token back into the scanner (where it actually never was observed).

Why is it invoked twice to begin with?